planner/cascades/optimize.go

```diff
@@ -183,6 +183,10 @@ func (opt *Optimizer) findMoreEquiv(g *memo.Group, elem *list.Element) (eraseCur
 		// the transformation rule to enumerate all the possible expressions.
 		iter := memo.NewExprIterFromGroupElem(elem, pattern)
 		for ; iter != nil && iter.Matched(); iter.Next() {
+			if iter.GetExpr().HasAppliedRule(rule) {
```

Comment:
Check this in `Match`?

Reply:
Agree. Then we only need this check for rules that can be added into the set.